backend/x11/backend.c

```diff
@@ -0,0 +1,374 @@
+		&input_impl, "X11 pointer", 0, 0);
+	wlr_pointer_init(&x11->pointer, &pointer_impl);
+	x11->pointer_dev.pointer = &x11->pointer;
+
+	return &x11->backend;
+
+error_event:
+	wl_event_source_remove(x11->event_source);
+error_x11:
+	xcb_disconnect(x11->xcb_conn);
+	XCloseDisplay(x11->xlib_conn);
+	free(x11);
+	return NULL;
+}
+
+static struct wlr_output_impl output_impl;
+
+static bool wlr_x11_backend_start(struct wlr_backend *backend) {
+	struct wlr_x11_backend *x11 = (struct wlr_x11_backend *)backend;
+	struct wlr_x11_output *output = &x11->output;
+
+	uint32_t mask = XCB_CW_BACK_PIXEL | XCB_CW_EVENT_MASK;
+	uint32_t values[2] = {
+		x11->screen->white_pixel,
+		XCB_EVENT_MASK_EXPOSURE |
+		XCB_EVENT_MASK_KEY_PRESS | XCB_EVENT_MASK_KEY_RELEASE |
+		XCB_EVENT_MASK_BUTTON_PRESS | XCB_EVENT_MASK_BUTTON_RELEASE |
+		XCB_EVENT_MASK_POINTER_MOTION |
+		XCB_EVENT_MASK_STRUCTURE_NOTIFY
+	};
+
+	output->x11 = x11;
+
+	wlr_output_init(&output->wlr_output, &output_impl);
+	snprintf(output->wlr_output.name, sizeof(output->wlr_output.name), "X11-1");
+
+	output->win = xcb_generate_id(x11->xcb_conn);
+	xcb_create_window(x11->xcb_conn, XCB_COPY_FROM_PARENT, output->win, x11->screen->root,
+		0, 0, 1024, 768, 1, XCB_WINDOW_CLASS_INPUT_OUTPUT,
+		x11->screen->root_visual, mask, values);
+
+	output->surf = wlr_egl_create_surface(&x11->egl, &output->win);
+	if (!output->surf) {
+		wlr_log(L_ERROR, "Failed to create EGL surface");
+		return false;
+	}
+
+	init_atom(x11, &x11->atoms.wm_protocols, 1, "WM_PROTOCOLS");
+	init_atom(x11, &x11->atoms.wm_delete_window, 0, "WM_DELETE_WINDOW");
+
+	xcb_change_property(x11->xcb_conn, XCB_PROP_MODE_REPLACE, output->win,
+		x11->atoms.wm_protocols.reply->atom, XCB_ATOM_ATOM, 32, 1,
+		&x11->atoms.wm_delete_window.reply->atom);
+
+	xcb_map_window(x11->xcb_conn, output->win);
+	xcb_flush(x11->xcb_conn);
+
+	wl_signal_emit(&x11->backend.events.output_add, output);
+	wl_signal_emit(&x11->backend.events.input_add, &x11->keyboard_dev);
+	wl_signal_emit(&x11->backend.events.input_add, &x11->pointer_dev);
+
+	wl_event_source_timer_update(x11->frame_timer, 16);
+
+	return true;
+}
+
+static void wlr_x11_backend_destroy(struct wlr_backend *backend) {
+	if (!backend) {
+		return;
+	}
+
+	struct wlr_x11_backend *x11 = (struct wlr_x11_backend *)backend;
+
+	wl_event_source_remove(x11->frame_timer);
+	wlr_egl_free(&x11->egl);
+
+	xcb_disconnect(x11->xcb_conn);
+	free(x11);
+}
+
+static struct wlr_egl *wlr_x11_backend_get_egl(struct wlr_backend *backend) {
+	struct wlr_x11_backend *x11 = (struct wlr_x11_backend *)backend;
+	return &x11->egl;
+}
+
+bool wlr_backend_is_x11(struct wlr_backend *backend) {
+	return backend->impl == &backend_impl;
+}
+
+static struct wlr_backend_impl backend_impl = {
+	.start = wlr_x11_backend_start,
+	.destroy = wlr_x11_backend_destroy,
+	.get_egl = wlr_x11_backend_get_egl,
+};
+
+static void output_transform(struct wlr_output *output, enum wl_output_transform transform) {
+	// TODO
```

Comment:
output_transform should be as easy as just doing this: https://github.com/SirCmpwn/wlroots/blob/master/backend/wayland/output.c#L51